File: kotlin/engine/src/main/kotlin/org/hpb/engine/Secp.kt
package org.hpb.engine

import fr.acinq.secp256k1.Secp256k1

/**
 * Thin wrapper over libsecp256k1 (ACINQ JNI bindings) for the operations the
 * engine needs: x-only key handling, BIP341 taproot output tweaking, and
 * deterministic per-job key derivation. Signing happens in Bitcoin Core.
 */
object Secp {
    private val impl = Secp256k1.get()

    fun publicKey(privkey: ByteArray): ByteArray = impl.pubkeyCreate(privkey)

    fun xonly(privkey: ByteArray): ByteArray = compress(publicKey(privkey)).copyOfRange(1, 33)

    fun xonlyHex(privkey: ByteArray): String = xonly(privkey).hex()

    private fun compress(pubkey: ByteArray): ByteArray = impl.pubKeyCompress(pubkey)

    /** Parse an x-only key as a curve point (even-Y lift); throws if invalid. */
    fun liftX(xonly: ByteArray): ByteArray {
        require(xonly.size == 32) { "x-only key must be 32 bytes" }
        return impl.pubkeyParse(byteArrayOf(0x02) + xonly)
    }

    fun validXonly(hex: String): ByteArray {
        require(hex.length == 64) { "expected 64-hex x-only pubkey" }
        val raw = hex.hexBytes()
        liftX(raw)
        return raw
    }

    /**
     * BIP341 output key: lift_x(internal) + H_TapTweak(internal || merkleRoot) * G.
     * Returns the output x-only key and its parity bit (for control blocks).
     */
    fun taprootOutput(internalXonly: ByteArray, merkleRoot: ByteArray): Pair<ByteArray, Int> {
        val tweak = taggedHash("TapTweak", internalXonly + merkleRoot)
        val tweaked = compress(impl.pubKeyTweakAdd(liftX(internalXonly), tweak))
        val parity = if (tweaked[0].toInt() == 0x03) 1 else 0
        return tweaked.copyOfRange(1, 33) to parity
    }

    /** Scalar addition mod n on a private key (per-job genesis derivation).
     *  The JNI binding mutates its input, so operate on a copy. */
    fun privateKeyTweakAdd(privkey: ByteArray, tweak: ByteArray): ByteArray =
        impl.privKeyTweakAdd(privkey.copyOf(), tweak)

    /** BIP340 Schnorr signature over a 32-byte message (Nostr event ids). */
    fun schnorrSign(message32: ByteArray, privkey: ByteArray): ByteArray =
        impl.signSchnorr(message32, privkey, null)

    fun schnorrVerify(signature: ByteArray, message32: ByteArray, xonly: ByteArray): Boolean =
        impl.verifySchnorr(signature, message32, xonly)

    /** Raw X coordinate of the ECDH shared point (NIP-44 needs it unhashed). */
    fun ecdhX(privkey: ByteArray, xonlyPeer: ByteArray): ByteArray {
        val shared = compress(impl.pubKeyTweakMul(liftX(xonlyPeer), privkey.copyOf()))
        return shared.copyOfRange(1, 33)
    }
}

File: ios/HpbCore/Sources/HpbCore/Secp.swift
import Foundation
import secp256k1

/// Thin wrapper over libsecp256k1 for the worker's needs: x-only keys,
/// BIP340 schnorr (Nostr event ids), and the raw ECDH x-coordinate NIP-44
/// wants. All library API use is isolated here.
public enum Secp {
    public static func xonly(_ privkey: [UInt8]) throws -> [UInt8] {
        try secp256k1.Signing.PrivateKey(dataRepresentation: privkey).publicKey.xonly.bytes
    }

    public static func xonlyHex(_ privkey: [UInt8]) throws -> String {
        try xonly(privkey).hex
    }

    /// Deterministic BIP340 signature: zeroed aux randomness is what
    /// libsecp uses for a NULL aux pointer, so signatures match the Kotlin
    /// reference (ACINQ passes null) byte-for-byte — the vector corpus
    /// depends on that.
    public static func schnorrSign(_ message32: [UInt8], _ privkey: [UInt8]) throws -> [UInt8] {
        var message = message32
        var aux = [UInt8](repeating: 0, count: 32)
        let key = try secp256k1.Schnorr.PrivateKey(dataRepresentation: privkey)
        return try Array(key.signature(message: &message, auxiliaryRand: &aux).dataRepresentation)
    }

    public static func schnorrVerify(
        _ signature: [UInt8],
        _ message32: [UInt8],
        _ xonly: [UInt8]
    ) -> Bool {
        guard signature.count == 64, message32.count == 32, xonly.count == 32 else { return false }
        guard let sig = try? secp256k1.Schnorr.SchnorrSignature(dataRepresentation: signature) else {
            return false
        }
        var message = message32
        return secp256k1.Schnorr.XonlyKey(dataRepresentation: xonly).isValid(sig, for: &message)
    }

    /// Raw X coordinate of the ECDH shared point (NIP-44 needs it
    /// unhashed); throws on an invalid peer key or an out-of-range privkey.
    public static func ecdhX(_ privkey: [UInt8], _ peerXonly: [UInt8]) throws -> [UInt8] {
        guard privkey.count == 32, peerXonly.count == 32 else {
            throw HpbError.invalid("bad key length")
        }
        var pubkey = secp256k1_pubkey()
        let compressed = [UInt8]([0x02] + peerXonly)
        guard secp256k1_ec_pubkey_parse(
            secp256k1.Context.rawRepresentation, &pubkey, compressed, compressed.count
        ) == 1 else {
            throw HpbError.invalid("invalid x-only key")
        }
        var shared = [UInt8](repeating: 0, count: 32)
        guard secp256k1_ecdh(
            secp256k1.Context.rawRepresentation, &shared, &pubkey, privkey,
            { output, x32, _, _ in
                memcpy(output, x32, 32)
                return 1
            }, nil
        ) == 1 else {
            throw HpbError.invalid("invalid private key")
        }
        return shared
    }
}
